docs/programming-oak.md

```diff
@@ -328,9 +328,86 @@ passes this on to the Oak Runtime, which will then notify the running Nodes that
+- Information flow analysis can be more precise and fine-grained if components
+  are smaller and the interactions between them are constrained.
+
+The first step in building a multi-Node Application is to write the code for all
+of the Nodes; the `ApplicationConfiguration` needs to include the configuration
+and code for any Node that might get run as part of the Application. New Node
+types cannot be added after the application starts; any Node that the
+Application might need has to be included in the original configuration.
+
+As before, each Node must include an `oak_main(u64) -> u32` entrypoint, but for
+an internal Node it's entirely up to the Application developer as to what
+channel handle gets passed to this entrypoint, and as to what messages are sent
+down that channel. The application may choose to use protobuf-encoded messages
+(as gRPC does) for its internal communications, or something else entirely (e.g.
+the [serde crate](https://crates.io/crates/serde)).
+
+Regardless of how the application communicates with the new Node, the typical
+pattern is to:
+
+- Create a new channel with the
+  [`channel_create`](https://project-oak.github.io/oak/sdk/oak/fn.channel_create.html)
+  host function, receiving local handles for both halves of the channel.
+- Create a new Node instance with the
+  [`node_create`](https://project-oak.github.io/oak/sdk/oak/fn.node_create.html)
+  host function, passing in the handle for the read half of the new channel.
+- Afterwards, close the local handle for the read half, as it is no longer
+  needed.
+
+For example, the example Chat application creates a Node for each chat room and
+saves off the write handle that will be used to send messages to the room:
 
-TODO: explore the primitives available here
+<!-- prettier-ignore-start -->
+[embedmd]:# (../examples/chat/module/rust/src/lib.rs Rust /.*Create a new Node/ /^ +\);$/)
+```Rust
+        // Create a new Node for this room.
+        let (wh, rh) = oak::channel_create().unwrap();
+        oak::node_create("room-config", rh);
+        oak::channel_close(rh.handle);
+
+        self.rooms.insert(
+            req.admin_id.clone(),
+            RoomRef {
+                name: req.name,
+                channel: wh,
+            },
+        );
+```
+<!-- prettier-ignore-end-->
+
+The same code (identified by `"room-config"`) will be run for each per-room
+Node, but each instance will have its own Web Assembly linear memory (≈heap) and
+stack.
+
+Once a new Node has started, the existing Node can communicate with the new Node
+by sending messages over the channel via `channel_write`. Of course, the new
+Node only has a handle to the read half of a channel, and so only has a way of
+_receiving_.
+
+To cope with this, it's normal for the inbound messages to be accompanied by a
+handle for the _write_ half of a different channel, which is then used for
+responses &ndash; so the new Node has a way of _sending_ externally, as well as
```

Comment:
(super nit)

```suggestion
responses &mdash; so the new Node has a way of _sending_ externally, as well as
```

Reply:
Robert Bringhurst, "The Elements of Typographic Style (version 2.5)", 2002, §5.2.1, p.80

> The em dash is the nineteenth century standard, still prescribed in many editorial style books, but the em dash is too long for use with the best text faces. Like the oversized space between sentences, it belongs to the padded and corseted aesthetic of Victorian typography.

(BTW, em dashes are usually set without spaces&mdash;like this.)